casper/src/main/resources/RevVault.rho

```diff
@@ -112,32 +112,35 @@ in {
               }
             } |
 
-            contract _getOrCreate(@revAddress, constructor, retCh) = {
-              new revAddressValidCh, revAddressValidEitherCh, getOrCreate in {
-                RevAddress!("validate", revAddress, *revAddressValidCh) |
-                @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
-                @Either!("flatMap <-", *revAddressValidEitherCh, *getOrCreate, *retCh) |
-                for (_, ret <- getOrCreate) {
-                  for (@vaultMap <- vaultMapStore) {
-                    match vaultMap.get(revAddress) {
-                      Nil => {
-                        new resCh in {
-                          constructor!(*resCh) |
-                          for (@eitherVault <- resCh) {
-                            match eitherVault {
-                              (true, vault) => vaultMapStore!(vaultMap.set(revAddress, vault))
-                              (false,  _    ) => vaultMapStore!(vaultMap)
-                            } |
-                            ret!(eitherVault)
-                          }
+            new _getOrCreateCombinator in {
+              contract _getOrCreateCombinator(@(revAddress, constructor), _, ret) = {
+                for (@vaultMap <- vaultMapStore) {
+                  match vaultMap.get(revAddress) {
+                    Nil => {
+                      new resCh in {
+                        @constructor!(*resCh) |
+                        for (@eitherVault <- resCh) {
+                          match eitherVault {
+                            (true, vault) => vaultMapStore!(vaultMap.set(revAddress, vault))
+                            (false,  _    ) => vaultMapStore!(vaultMap)
+                          } |
+                          ret!(eitherVault)
                         }
                       }
-                      value => {
-                        vaultMapStore!(vaultMap) | ret!((true, value))
-                      }
+                    }
+                    value => {
+                      vaultMapStore!(vaultMap) | ret!((true, value))
                     }
                   }
                 }
+              } |
+
+              contract _getOrCreate(@revAddress, constructor, retCh) = {
+                new revAddressValidCh, revAddressValidEitherCh in {
+                  RevAddress!("validate", revAddress, *revAddressValidCh) |
+                  @Either!("fromNillableError <-", *revAddressValidCh, *revAddressValidEitherCh) |
+                  @Either!("flatMapClean <-", *revAddressValidEitherCh, *_getOrCreateCombinator, (revAddress, *constructor), *retCh)
```

Comment:
I have big issues with how the end-user API changed in the previous PR (sorry for missing its review).  This method lost resemblence to flatMap to the point where it's really confusing what does what. Why does the combinator ignore its arguments?

I really don't feel comfortable merging these changes without having the benchmarks showing that self-respawning linear receive is indeed slower.

Reply:
Since all of the combinators are supposed to be stateless (are they?), we should be able to put them in the top-level scope and never declare another copy of them past genesis.